// rooks/dist/esm/hooks/useMouse.js
import { useState, useEffect } from 'react';

const initialMouseState = {
    clientX: null,
    clientY: null,
    movementX: null,
    movementY: null,
    offsetX: null,
    offsetY: null,
    pageX: null,
    pageY: null,
    screenX: null,
    screenY: null,
    x: null,
    y: null,
};
function getMousePositionFromEvent(event) {
    const { screenX, screenY, movementX, movementY, pageX, pageY, clientX, clientY, offsetX, offsetY, } = event;
    return {
        clientX,
        clientY,
        movementX,
        movementY,
        offsetX,
        offsetY,
        pageX,
        pageY,
        screenX,
        screenY,
        x: screenX,
        y: screenY,
    };
}
/**
 * useMouse hook
 *
 * Retrieves current mouse position and information about the position like
 * screenX, pageX, clientX, movementX, offsetX
 * @see https://rooks.vercel.app/docs/useMouse
 */
function useMouse() {
    const [mousePosition, setMousePosition] = useState(initialMouseState);
    function updateMousePosition(event) {
        setMousePosition(getMousePositionFromEvent(event));
    }
    useEffect(() => {
        document.addEventListener("mousemove", updateMousePosition);
        return () => {
            document.removeEventListener("mousemove", updateMousePosition);
        };
    }, []);
    return mousePosition;
}

export { useMouse };
